File: encryption/c-sharp/caesar_cipher.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;
using System.Collections.Generic;

namespace IsaacCodeSamples
{
    class CaesarCipher
    {
        readonly static char[] alphabet = { //Read-only as we will not be changing the values of the array
            'a', 'b', 'c', 'd', 'e', 'f',
            'g', 'h', 'i', 'j', 'k', 'l',
            'm', 'n', 'o', 'p', 'q', 'r',
            's', 't', 'u', 'v', 'w', 'x',
            'y', 'z'
        };
        public static void Main()
        {
            string plaintext = "a byte is eight bits";
            int key = 13;

            string ciphertext = CaesarShift(plaintext, key);

            Console.WriteLine($"\nPlaintext: {plaintext}");
            Console.WriteLine($"\nCiphertext: {ciphertext}");

            string decrypt = CaesarShift(ciphertext, -key);
            Console.WriteLine($"\nDecrypted ciphertext: {decrypt}");
        }

        public static string CaesarShift(string plaintext, int key)
        {
            string ciphertext = "";
            foreach (char c in plaintext.ToCharArray())
            {
                if (c != ' ')
                {
                    int this_index = Array.IndexOf(alphabet, c);
                    int new_index = (this_index += key) % alphabet.Length; // Wrap around to the start of the alphabet if new index is past z
                    if (new_index < 0)
                    {
                        new_index += alphabet.Length; // Avoid problems with negative indices.
                    }
                    ciphertext += alphabet[new_index].ToString();
                }
                else
                {
                    ciphertext += ' '.ToString();
                }
            }
            return ciphertext;
        }
    }
}

File: encryption/python/caesar-cipher.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4

import string

"""
    The Caesar Cipher uses a constant integer: "key" to convert plaintext to ciphertext
    by shifting each letter along in the alphabet by the value of the key.
"""


def caesar_cipher_shift(plaintext, key):
    # This shift will only work on lowercase letters!
    alphabet = string.ascii_lowercase
    ciphertext = ""

    for letter in plaintext:
        if letter == " ":
            ciphertext += " "
        else:
            current_index = alphabet.index(letter)
            shifted_index = (current_index + key) % len(alphabet)
            ciphertext += alphabet[shifted_index]

    return ciphertext


def main():
    plaintext = "a byte is eight bits"
    key = 13

    print(f"\nPlaintext: {plaintext}")

    print("\n--- Encrypting ---")
    encrypt = caesar_cipher_shift(plaintext, key)
    print(encrypt)

    print("\n--- Decrypting ---")
    decrypt = caesar_cipher_shift(encrypt, -key)
    print(decrypt)


# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == "__main__":
    main()
